### DISCUSSION TITLE
Trigger Placeholders

### DISCUSSION THREAD

**Discussion by AlessioGr** [PlaceholderAPI/PlaceholderAPI#745] [(3) Ideas]
### Confirmation

- [X] I checked the Issues and Pull request tab for any existing issues or PRs.
- [X] My Feature Request is for **PlaceholderAPI** and not any expansion or other plugin.

### Type

New API feature

### Description

Currently, I can read PlaceholderAPI placeholders by “pulling” the information from the plugin which adds these placeholders. If a continuously changing placeholder should be displayed somewhere, it would need to be updated/requested in a set interval.

This introduces (1) an **update delay**, and (2) **performance issues** when it has to be updated even if the value of the placeholder didn't change.

A better solution for these placeholders: **Trigger placeholders**. So, the value of the placeholder is “pushed” to PlaceholderAPI every time it changes. Plugins can listen to that and update it whenever the value changes.

**One of many useful scenarios:**
PVP Plugin. The PVP status is displayed in Tab via a pvpstatus player placeholder. Each player can change their PVP status with a toggle command. With the current system, there is an update delay and the placeholder gets refreshed way too often — the PVP status only changes rarely.

With a trigger placeholder system, the change could be reflected immediately, and it would be better for the performance. The plugin author could push the placeholder update right in their hypothetical player.setPvPStatus() method.

### Additional Information

Could be hard to implement in PlaceholderAPI. But it would be an addition/option and thus wouldn't affect existing servers.

Still worth mentioning that feature request, even if it's difficult to implement.

**Comment by Andre601** [comment]
The biggest issue I see with this is, that placeholders are rarely static and may change a lot. Not to mention that they can and often will be per-player.

Just take the `%player_name%` placeholder as an example.
Even if we would keep a sort of cache could this end up in caching hundreds of entries because a server could have hundreds of players online and either a chat, tab or scoreboard plugin could then call the `setPlaceholders` method of PlaceholderAPI frequently here.

Other placeholders that f.e. would return coordinates would also make the caching complicated as we would need to include a system to keep track of past values, check if they have changed and then return the new or cached value based on this.

Such a mechanic would for sure bloat PlaceholderAPI and make it consume much more processing power than currently needed.
It should imo be the job of the actual expansions to keep a cache of recently returned values to increase the speed of returning placeholders.

Making PlaceholderAPI cache the request would sure not improve performance and speed as, due to its current setup, PlaceholderAPI would still need to parse the String to see if it changed compared to the caching.

For the sake of understanding did I make this flowchart to show how PlaceholderAPI works when a plugin requests placeholders to be parsed using the `setPlaceholders` method:
[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBwbHVnaW5bUGx1Z2luXSAtLT58Q2FsbCBzZXRQbGFjZWhvbGRlcnN8IHBhcGlbUGxhY2Vob2xkZXJBUEldXG4gICAgcGFwaSAtLT58Q2hlY2sgZm9yIHBsYWNlaG9sZGVyc3wgZm91bmRQbGFjZWhvbGRlcntQbGFjZWhvbGRlcnMgZm91bmQ_fVxuICAgIGZvdW5kUGxhY2Vob2xkZXIgLS0-IHBsYWNlaG9sZGVyWWVzKFllcylcbiAgICBmb3VuZFBsYWNlaG9sZGVyIC0tPiBwbGFjZWhvbGRlck5vKE5vKVxuXG4gICAgcGxhY2Vob2xkZXJObyAtLT58UmV0dXJuIHVucGFyc2VkIFN0cmluZ3wgcGx1Z2luXG5cbiAgICBwbGFjZWhvbGRlclllcyAtLT58ZmluZCBQbGFjZWhvbGRlckV4cGFuc2lvbnwgZXhwYW5zaW9uRm91bmR7UGxhY2Vob2xkZXJFeHBhbnNpb24gZm91bmQ_fVxuICAgIGV4cGFuc2lvbkZvdW5kIC0tPiBleHBhbnNpb25ZZXMoWWVzKVxuICAgIGV4cGFuc2lvbkZvdW5kIC0tPiBleHBhbnNpb25ObyhObylcblxuICAgIGV4cGFuc2lvblllcyAtLT58Q2FsbCBvblJlcXVlc3R8IGV4cGFuc2lvbltQbGFjZWhvbGRlckV4cGFuc2lvbl1cblxuICAgIGV4cGFuc2lvbiAtLT58UmV0dXJuIHBhcnNlZCBwbGFjZWhvbGRlcnwgbW9yZVBsYWNlaG9sZGVyc3tNb3JlIHBsYWNlaG9sZGVycz99XG4gICAgbW9yZVBsYWNlaG9sZGVycyAtLT4gcGxhY2Vob2xkZXJZZXNcbiAgICBtb3JlUGxhY2Vob2xkZXJzIC0tPiBtb3JlTm8oTm8pXG5cbiAgICBtb3JlTm8gLS0-fFJldHVybiBwYXJzZWQgU3RyaW5nfCBwbHVnaW5cblxuICAgIGV4cGFuc2lvbk5vIC0tPiBtb3JlUGxhY2Vob2xkZXJzIiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQiLCJ0ZXN0IjoiaGVsbG8gd29ybGQifSwidXBkYXRlRWRpdG9yIjp0cnVlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6dHJ1ZSwibG9hZGVyIjp7InR5cGUiOiJnaXN0IiwiY29uZmlnIjp7InVybCI6Imh0dHBzOi8vZ2lzdC5naXRodWIuY29tLzYyNjhhMjNlNjczYTUzM2RjYjE5OGYyNDFmZDcwMTJhLzRlYjAzODg3ZTZhNDEzOTdlODBiZGNkYmY5NDAxN2M0OThmOGYxZTIifX19)](https://mermaid-js.github.io/mermaid-live-editor/beta/edit/#)

Please keep in mind that this is only my personal opinion and that the actual maintainers may have their own view on this idea, so don't be discouraged by me not thinking it would be beneficial.

Also note that for PlaceholderAPI V3, one suggested change is asynchronous placeholder parsing/handling which could improve the overall performance of PlaceholderAPI, but there is currently no guarantee of this being implemented in the near future.
